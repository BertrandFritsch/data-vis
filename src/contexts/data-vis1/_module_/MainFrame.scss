
.data-vis-container {
  display: flex;
  flex-direction: column;
  height: 100%;
  width: 1000px;
  margin: 0 auto;

  .data-vis-header {
    margin-top: 20px;
    display: flex;
    flex-direction: column;
    p {
      text-align: center;
    }

    .data-vis-header-1 {
      font-size: 2rem;
    }

    .data-vis-header-2 {
      font-size: 1.5rem;
    }
  }

  .data-vis-body {
    flex: 1;
    display: flex;
    min-height: 500px;

    .data-vis-chart {
      flex: 1;
      align-self: stretch;
      margin: 100px 0 100px 100px;
      min-width: 100px;

      svg {

        .data-vis-line {
          fill: none;
          stroke-linejoin: round;
          stroke-linecap: round;
          stroke-width: 1.5;

          &.data-vis-line-over {
            stroke-width: 2.5;
          }
        }

        .data-vis-axis-y {
          .domain {
            display: none;
          }
          .tick line {
            stroke: #767676;
          }

          .tick:not(.data-vis-axis-y-zero):not(:first-of-type) line {
            stroke-dasharray: 2, 2;
          }

          .data-vis-axis-y-zero line {
            stroke-width: 2;
          }
        }

        text {
          font-size: .75rem;
          font-family: ibm-plex-sans;
        }
      }
    }

    .data-vis-tools {
      align-self: center;

      margin: 100px 50px;
      display: flex;
      flex-direction: column;

      .bx--toggle__label {
        width: 10rem;
        margin: .5rem;
        font-size: .875rem;

        .bx--toggle__appearance {
          font-size: .875rem;
          margin-right: .5rem;
        }
      }

      .data-vis-container-place-over {
        font-weight: 600;
      }
    }
  }

  .data-vis-footer {
    display: flex;
    flex-direction: column;
    margin: 10px;
    p {
      text-align: center;
      font-size: .75rem;
    }
  }
}
