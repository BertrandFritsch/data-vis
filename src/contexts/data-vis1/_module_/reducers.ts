import { MainFrameAction } from './actionTypes';

// tslint:disable-next-line no-var-requires
const data: RawRecord[] = require('../../../../data/data.json');

export type WorldPlace =
  'Glob'
  | 'NHem'
  | 'SHem'
  | '24N-90N'
  | '24S-24N'
  | '90S-24S'
  | '64N-90N'
  | '44N-64N'
  | '24N-44N'
  | 'EQU-24N'
  | '24S-EQU'
  | '44S-24S'
  | '64S-44S'
  | '90S-64S';

export type PlaceProperties = {
  [ key in WorldPlace ]: {
    order: number;
    color: string;
    text: string;
  }
};

export const placeProperties: PlaceProperties = {
  'Glob':    { order: 0, color: '#ff0000', text: 'Global' },
  'NHem':    { order: 1, color: '#0391f9', text: 'North hemisphere' },
  '24N-90N': { order: 2, color: '#5b924e', text: '24N-90N' },
  '64N-90N': { order: 3, color: '#fc9127', text: '64N-90N' },
  '44N-64N': { order: 4, color: '#ab5fab', text: '44N-64N' },
  '24N-44N': { order: 5, color: '#3bbbce', text: '24N-44N' },
  'EQU-24N': { order: 6, color: '#0712be', text: 'EQU-24N' },
  '24S-24N': { order: 7, color: '#cc4c76', text: '24S-24N' },
  '24S-EQU': { order: 8, color: '#5a5b12', text: '24S-EQU' },
  '44S-24S': { order: 9, color: '#416920', text: '44S-24S' },
  '64S-44S': { order: 10, color: '#f25d22', text: '64S-44S' },
  '90S-64S': { order: 11, color: '#3e587a', text: '90S-64S' },
  '90S-24S': { order: 12, color: '#7c213e', text: '90S-24S' },
  'SHem':    { order: 13, color: '#310035', text: 'South hemisphere' }
};

export type RawRecord = {
  [ key in WorldPlace ]: number | null;
  } & { Year: number };

export interface TemperatureByDate {
  date: Date;
  temperature: number;
}

export interface TemperaturesByPlace {
  place: WorldPlace;
  values: TemperatureByDate[];
}

export interface State {
  data: RawRecord[];
  placeProperties: PlaceProperties;
}

const initialState: State = {
  data,
  placeProperties
};

export interface MainFrameState {
  mainFrame: State;
}

const reducer = (state: State = initialState, action: MainFrameAction): State => {
  switch (action.type) {

    default:
      return state;
  }
};

export default { mainFrame: reducer };
