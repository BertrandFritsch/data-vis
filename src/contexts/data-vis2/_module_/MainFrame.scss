
.data-vis-container {
  display: flex;
  flex-direction: column;
  height: 100%;
  width: 1000px;
  margin: 0 auto;

  .data-vis-header {
    margin-top: 20px;
    display: flex;
    flex-direction: column;
    p {
      text-align: center;
    }

    .data-vis-header-1 {
      font-size: 2rem;
    }

    .data-vis-header-2 {
      font-size: 1.5rem;
    }
  }

  .data-vis-body {
    flex: 1;
    display: flex;
    height: 500px;
    width: 500px;
    align-self: center;

    .data-vis-chart {
      flex: 1;
      align-self: stretch;
      margin: 20px;

      svg {

        circle {
          fill: #6898cf;

          &:hover {
            fill: #5b7faf;
          }
        }

        text {
          font-size: .2rem;
          font-family: ibm-plex-sans;
          text-anchor: middle;
          dominant-baseline: central;
          fill: white;
          pointer-events: none;
        }
      }
    }

    .data-vis-tooltip {
      opacity: 0;
      position: absolute;
      pointer-events: none;
      transition: opacity 300ms;

      &.data-vis-tooltip--hover {
        opacity: .9;
      }

      div {
        font-weight: 100;
        font-size: .75rem;
        background-color: black;
        color: white;
        text-align: left;
        padding: 10px;
        border: 0;
        border-radius: 2px;

        span {
          line-height: 1.25rem;
        }
      }
    }
  }

  .data-vis-footer {
    display: flex;
    flex-direction: column;
    margin: 10px;
    p {
      text-align: center;
      font-size: .75rem;
    }
  }
}
