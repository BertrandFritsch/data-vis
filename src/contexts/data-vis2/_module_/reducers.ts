import { MainFrameAction } from './actionTypes';

// tslint:disable-next-line no-var-requires
const data: RawRecord[] = require('../../../../data/data2.json');

export interface RawRecord {
  geo: string;
  population: number;
}

export interface State {
  data: RawRecord[];
}

const initialState: State = {
  data
};

export interface MainFrameState {
  mainFrame: State;
}

const reducer = (state: State = initialState, action: MainFrameAction): State => {
  switch (action.type) {

    default:
      return state;
  }
};

export default { mainFrame: reducer };
