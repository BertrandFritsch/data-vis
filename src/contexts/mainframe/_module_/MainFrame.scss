
.data-vis-container {
  display: flex;
  flex-direction: column;
  height: 100%;

  .data-vis-header {
    height: 50px;
  }

  .data-vis-body {
    flex: 1;
    padding: 100px;

    svg {
      height: 100%;
      width: 100%;
    }

  }
}