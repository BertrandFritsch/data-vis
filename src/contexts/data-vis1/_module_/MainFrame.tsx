import * as React from 'react';
import './MainFrame.scss';
import * as d3 from 'd3';
import { PlaceProperties, TemperatureByDate, TemperaturesByPlace, WorldPlace } from './reducers';
import { setInArray } from '../../../helpers';

export interface Props {
  data: TemperaturesByPlace[];
  minValue: number;
  maxValue: number;
  minDate: Date;
  maxDate: Date;
  placeProperties: PlaceProperties;
}

interface State {
  selectedPlaces: WorldPlace[];
}

interface ChartState {
  update: () => void;
  resize: () => void;
}

export default class MainFrame extends React.PureComponent<Props, State> {
  state: State;
  private chartState: ChartState = { update: () => undefined, resize: () => undefined };

  constructor(props: Props) {
    super(props);

    this.state = { selectedPlaces: props.data.find(d => d.place === 'Glob') ? [ 'Glob' ] : [] };
  }

  componentDidMount() {
    window.addEventListener('resize', this.windowResized);
  }

  componentDidUpdate() {
    this.chartState.update();
  }

  componentWillUnmount() {
    window.removeEventListener('resize', this.windowResized);
  }

  render() {
    return (
      <div className='data-vis-container'>
        <div className='data-vis-header'>
          <p className='data-vis-header-1'>Zonal Annual Mean Land-Ocean Temperature</p>
          <p className='data-vis-header-2'>{ `${ this.props.minDate.getFullYear() } - ${ this.props.maxDate.getFullYear() }` }</p>
        </div>
        <div className='data-vis-body'>
          <div className='data-vis-chart' tabIndex={ -1 } ref={ this.renderChart } />
          <div className='data-vis-tools'>
            {
              this.props.data.map(r =>
                <div key={ r.place } style={ { color: this.props.placeProperties [ r.place ].color } }>
                  <div className='bx--form-item'>
                    <input className='bx--toggle bx--toggle--small' id={ `toggle-${ r.place }` } type='checkbox'
                           aria-label='Label Name'
                           checked={ this.state.selectedPlaces.find(d => d === r.place) !== undefined }
                           onChange={ e => this.togglePlace(r.place, e.target.checked) }
                    />
                    <label className='bx--toggle__label' htmlFor={ `toggle-${ r.place }` }>
                      <span className='bx--toggle__appearance'>
                      </span>
                      { this.props.placeProperties [ r.place ].text }
                    </label>
                  </div>
                </div>
              )
            }
          </div>
        </div>
        <div className='data-vis-footer'>
          <p className='data-vis-footer-1'>GISTEMP Team, 2018: GISS Surface Temperature Analysis (GISTEMP). NASA Goddard
            Institute for Space Studies. Dataset accessed 2018-02-10 at https://data.giss.nasa.gov/gistemp/</p>
          <p className='data-vis-footer-2'>Hansen, J., R. Ruedy, M. Sato, and K. Lo, 2010: Global surface temperature
            change, Rev. Geophys., 48, RG4004, doi:10.1029/2010RG000345</p>
        </div>
      </div>
    );
  }

  private windowResized = () => this.chartState.resize();

  private renderChart = (elem: HTMLElement | null) => {
    if (elem !== null) {
      this.chartState = (() => {

        let clientRect = elem.getBoundingClientRect();
        const margin = { top: 20, right: 40, bottom: 20, left: 40 };

        const $svg = d3.select(elem)
                       .append<SVGSVGElement>('svg')
                       .attr('width', clientRect.width)
                       .attr('height', clientRect.height);

        const startDate = this.props.data[ 0 ].values[ 0 ].date;
        const endDate = this.props.data[ 0 ].values[ this.props.data[ 0 ].values.length - 1 ].date;

        const $gRoot = $svg.append<SVGGElement>('g')
                           .attr('transform', `translate(${ margin.left },${ margin.top })`);

        const x = d3.scaleTime()
                    .domain([ startDate, endDate ]);

        const y = d3.scaleLinear()
                    .domain([ this.props.minValue, this.props.maxValue ])
                    .nice();

        const line = d3.line<TemperatureByDate>()
                       .x(d => x(d.date))
                       .y(d => y(d.temperature));

        const nbXTicks = (nbIntervals: number) => {
          // replace the last automatic tick by the real last one
          const ticks = x.ticks(nbIntervals);
          ticks[ ticks.length - 1 ] = endDate;
          return ticks;
        };

        const xAxis = d3.axisBottom(x);

        $gRoot.append<SVGGElement>('g')
              .attr('class', 'data-vis-axis-x');

        const yFormat = d3.format('.1f');

        const yAxis = d3.axisLeft<number>(y)
                        .tickFormat(d => d === 0 ? '0' : `${ yFormat(d) }Â°`);

        $gRoot.append<SVGGElement>('g')
              .attr('class', 'data-vis-axis-y')
              .call(yAxis);

        $gRoot.append('text')
              .attr('class', 'data-vis-info')
              .attr('dy', -3)
              .text('Annual temperature change compared to the mean of the years from 1951 to 1980');

        return {
          update: () => {

            const width = clientRect.width - margin.left - margin.right;
            const height = clientRect.height - margin.top - margin.bottom;
            const nbIntervals = Math.max(Math.trunc(width / 50), 1);

            $svg.attr('width', clientRect.width)
                .attr('height', clientRect.height);

            x.rangeRound([ 0, width ]);
            y.rangeRound([ height, 0 ]);

            xAxis.tickValues(nbXTicks(nbIntervals));
            yAxis.tickSize(-(width + 20));

            $gRoot.select<SVGGElement>('.data-vis-axis-x')
                  .transition()
                  .attr('transform', `translate(0,${ height })`)
                  .call(xAxis);

            $gRoot.select<SVGGElement>('.data-vis-axis-y')
                  .transition()
                  .call(yAxis)
                  .call($g => {
                    $g.selectAll('.tick')
                      .attr('class', d => `tick ${ d === 0 ? 'data-vis-axis-y-zero' : '' }`)
                      .call($g1 => $g1.select('line')
                                      .attr('x1', -20))
                      .call($g1 => $g1.select('text')
                                      .attr('dx', -3)
                                      .attr('dy', -3));
                  });

            const $chartLines = $gRoot.selectAll<SVGGElement, TemperaturesByPlace>('.data-vis-group-line')
                                      .data(this.props.data.filter(d => this.state.selectedPlaces.find(p => p === d.place) !== undefined), d => d.place);

            $chartLines.exit()
                       .remove();

            $chartLines.enter()
                       .append<SVGGElement>('g')
                       .attr('class', 'data-vis-group-line')
                       .call($g => $g.append<SVGPathElement>('path')
                                     .attr('class', 'data-vis-line')
                                     .style('stroke', d => this.props.placeProperties [ d.place ].color)
                       )
                       .merge($chartLines)
                       .transition()
                       .call($g => $g.select('.data-vis-line')
                                     .attr('d', d => line(d.values))
                       );
          },

          resize: () => {
            clientRect = elem.getBoundingClientRect();
            this.chartState.update();
          }
        };
      })();
      this.chartState.update();
    }
  };

  private togglePlace(place: WorldPlace, checked: boolean) {
    this.setState({
      selectedPlaces: setInArray(this.state.selectedPlaces, p => p === place, checked ? place : undefined)
    });
  }
}
