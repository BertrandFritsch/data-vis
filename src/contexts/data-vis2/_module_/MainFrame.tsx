import * as React from 'react';
import './MainFrame.scss';
import * as d3 from 'd3';
import { Record } from './reducers';

export interface Props {
  data: Record[];
}

interface Node {
  data: Record;
  x: number;
  y: number;
}

interface State {
  nodes: Node[];
}

interface ChartState {
  update: () => void;
}

export default class MainFrame extends React.PureComponent<Props, State> {
  state: State;

  private chartState: ChartState = { update: () => undefined };

  constructor(props: Props) {
    super(props);

    this.state = {
      nodes: props.data.map(d => ({
        data: d,
        x: 50,
        y: 50
      }))
    };
  }

  render() {
    return (
      <div className='data-vis-container'>
        <div className='data-vis-header'>
          <p className='data-vis-header-1'>European Union Population</p>
          <p className='data-vis-header-2'>2017</p>
        </div>
        <div className='data-vis-body'>
          <div className='data-vis-chart' tabIndex={ -1 } ref={ this.renderChart } />
        </div>
        <div className='data-vis-footer'>
          <p className='data-vis-footer-1'>Source: https://ec.europa.eu/eurostat/</p>
          <p className='data-vis-footer-2'>Copyright: https://ec.europa.eu/info/legal-notice_en</p>
        </div>
      </div>
    );
  }

  private renderChart = (elem: HTMLElement | null) => {
    if (elem !== null) {
      this.chartState = (() => {

        const clientRect = elem.getBoundingClientRect();

        const scale = d3.scaleLinear()
                        .domain([ 0, 1 ])
                        .range([ 1, 20 ]);

        const $svg = d3.select(elem)
                       .append<SVGSVGElement>('svg')
                       .attr('width', clientRect.width)
                       .attr('height', clientRect.height)
                       .attr('viewBox', '0, 0, 100, 100');

        const $tooltipDiv = d3.select(elem)
                              .append('div')
                              .attr('class', 'data-vis-tooltip');

        $tooltipDiv.append('div');

        const $overallGroup = $svg.append<SVGGElement>('g');
        const $circlesData = $overallGroup.selectAll<SVGGElement, Node>('.data-vis-circle')
                                          .data(this.state.nodes, d => d.data.geo);

        const populationFormat = d3.format(',.3r');

        const $circleGroups = $circlesData.enter()
                                          .append<SVGGElement>('g')
                                          .attr('class', 'data-vis-group-circle')
                                          .each((d, i, g) => {
                                            const $g = d3.select(g [ i ]);
                                            const $circle = $g.append('circle')
                                                              .attr('r', scale(d.data.relativePopulation))
                                                              .on('mousemove', () => {
                                                                $tooltipDiv.classed('data-vis-tooltip--hover', true)
                                                                           .style('left', d3.event.pageX + 4 + 'px')
                                                                           .style('top', d3.event.pageY + 4 + 'px')
                                                                           .select('div')
                                                                           .html(`<span>${ d.data.geo }</span><br /><span>${ populationFormat(d.data.population)}</span>`);
                                                              })
                                                              .on('mouseout', () => {
                                                                $tooltipDiv.classed('data-vis-tooltip--hover', false);
                                                              });

                                            if (d.data.relativePopulation > .4) {
                                              $g.append('text')
                                                .text(d.data.geo);
                                            }
                                            else {
                                              $circle.append('title')
                                                     .text(d.data.geo);
                                            }
                                          })
                                          .merge($circlesData);

        d3.forceSimulation(this.state.nodes)
          .force('collide', d3.forceCollide<Node>().radius(d => scale(d.data.relativePopulation * 1.05)).strength(.75))
          .force('x', d3.forceX(50))
          .force('y', d3.forceY(50))
          .on('tick', () => {
            $circleGroups.each(
              (d, i, g) => {
                d.x = Math.min(Math.max(scale(d.data.relativePopulation), d.x), 100 - scale(d.data.relativePopulation));
                d.y = Math.min(Math.max(scale(d.data.relativePopulation), d.y), 100 - scale(d.data.relativePopulation));
                d3.select(g[ i ])
                  .attr('transform', `translate(${ d.x },${ d.y })`);
              }
            );
          });
        return {
          update: () => {
            // nothing to update
          }
        };
      })();
      this.chartState.update();
    }
  };
}
