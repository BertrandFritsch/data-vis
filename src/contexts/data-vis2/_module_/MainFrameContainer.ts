import { connect } from 'react-redux';
import MainFrame from './MainFrame';
import { MainFrameState, RawRecord } from './reducers';
import { createSelector } from 'reselect';

export function transformData(data: RawRecord[]) {
  const min = Math.min(...data.map(d => d.population));
  const max = Math.max(...data.map(d => d.population));

  return data.map(d => ({
    ...d,
    relativePopulation: (d.population - min) / (max - min)
  }));
}

const getData = createSelector(
  (state: MainFrameState) => state.mainFrame.data,
  data => transformData(data)
);

export default connect(
  (state: MainFrameState) => ({
    data: getData(state)
  })
)(MainFrame);
