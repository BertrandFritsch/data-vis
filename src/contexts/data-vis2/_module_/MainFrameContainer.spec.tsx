import { mount } from 'enzyme';
import * as React from 'react';
import { Provider } from 'react-redux';
import { createStore, Store } from 'redux';
import { MainFrameState } from './reducers';

// tslint:disable-next-line no-var-requires
const data = require('../../../../data/data2.json');

const MainFrame = (props: MainFrameProps) =>
  <div>{ props.data.map((n, i) => <p key={ i }>{ n.geo }</p>) }</div>;

jest.mock('./MainFrame', () => ({ default: MainFrame }));

import { Props as MainFrameProps } from './MainFrame';
import MainFrameContainer from './MainFrameContainer';

describe('MainFrameContainer', () => {

  const initialState = {
    data
  };

  const store: Store<MainFrameState> = createStore(() => ({ mainFrame: initialState }));

  it('should render correctly', () => {
    const wrapper = mount(
      <Provider store={ store }>
        <MainFrameContainer />
      </Provider>,
    );

    expect(wrapper).toMatchSnapshot();
  });
});
