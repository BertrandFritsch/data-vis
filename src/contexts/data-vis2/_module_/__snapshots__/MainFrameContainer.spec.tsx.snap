// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainFrameContainer should render correctly 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <Connect(MainFrame)>
    <MainFrame
      data={
        Array [
          Object {
            "geo": "Austria",
            "population": 8772865,
            "relativePopulation": 0.10117139153973453,
          },
          Object {
            "geo": "Belgium",
            "population": 11365834,
            "relativePopulation": 0.13265491063108673,
          },
          Object {
            "geo": "Bulgaria",
            "population": 7101859,
            "relativePopulation": 0.08088223678980731,
          },
          Object {
            "geo": "Croatia",
            "population": 4154213,
            "relativePopulation": 0.045092272036835745,
          },
          Object {
            "geo": "Cyprus",
            "population": 854802,
            "relativePopulation": 0.00503121877753437,
          },
          Object {
            "geo": "Czech Republic",
            "population": 10578820,
            "relativePopulation": 0.12309908088759136,
          },
          Object {
            "geo": "Denmark",
            "population": 5748769,
            "relativePopulation": 0.06445318004160949,
          },
          Object {
            "geo": "Estonia",
            "population": 1315635,
            "relativePopulation": 0.010626597879029646,
          },
          Object {
            "geo": "Finland",
            "population": 5503297,
            "relativePopulation": 0.06147268841274991,
          },
          Object {
            "geo": "France",
            "population": 67024459,
            "relativePopulation": 0.8084552702905784,
          },
          Object {
            "geo": "Germany",
            "population": 82800000,
            "relativePopulation": 1,
          },
          Object {
            "geo": "Greece",
            "population": 10757293,
            "relativePopulation": 0.12526607868154527,
          },
          Object {
            "geo": "Hungary",
            "population": 9797561,
            "relativePopulation": 0.11361312766493782,
          },
          Object {
            "geo": "Ireland",
            "population": 4774833,
            "relativePopulation": 0.052627765757923425,
          },
          Object {
            "geo": "Italy",
            "population": 60589445,
            "relativePopulation": 0.7303221008920554,
          },
          Object {
            "geo": "Latvia",
            "population": 1950116,
            "relativePopulation": 0.018330390202270005,
          },
          Object {
            "geo": "Lithuania",
            "population": 2847904,
            "relativePopulation": 0.029231224588638256,
          },
          Object {
            "geo": "Luxembourg",
            "population": 590667,
            "relativePopulation": 0.0018241232375590319,
          },
          Object {
            "geo": "Malta",
            "population": 440433,
            "relativePopulation": 0,
          },
          Object {
            "geo": "Netherlands",
            "population": 17081507,
            "relativePopulation": 0.20205392774830883,
          },
          Object {
            "geo": "Poland",
            "population": 37972964,
            "relativePopulation": 0.4557154969015318,
          },
          Object {
            "geo": "Portugal",
            "population": 10309573,
            "relativePopulation": 0.11982991605577528,
          },
          Object {
            "geo": "Romania",
            "population": 19638309,
            "relativePopulation": 0.2330983114566399,
          },
          Object {
            "geo": "Slovakia",
            "population": 5435343,
            "relativePopulation": 0.06064759908220498,
          },
          Object {
            "geo": "Slovenia",
            "population": 2065895,
            "relativePopulation": 0.019736164955796817,
          },
          Object {
            "geo": "Spain",
            "population": 46528966,
            "relativePopulation": 0.5596014486088787,
          },
          Object {
            "geo": "Sweden",
            "population": 9995153,
            "relativePopulation": 0.11601226606740174,
          },
          Object {
            "geo": "United Kingdom",
            "population": 65808573,
            "relativePopulation": 0.7936921280802751,
          },
        ]
      }
      dispatch={[Function]}
    >
      <div>
        <p
          key="0"
        >
          Austria
        </p>
        <p
          key="1"
        >
          Belgium
        </p>
        <p
          key="2"
        >
          Bulgaria
        </p>
        <p
          key="3"
        >
          Croatia
        </p>
        <p
          key="4"
        >
          Cyprus
        </p>
        <p
          key="5"
        >
          Czech Republic
        </p>
        <p
          key="6"
        >
          Denmark
        </p>
        <p
          key="7"
        >
          Estonia
        </p>
        <p
          key="8"
        >
          Finland
        </p>
        <p
          key="9"
        >
          France
        </p>
        <p
          key="10"
        >
          Germany
        </p>
        <p
          key="11"
        >
          Greece
        </p>
        <p
          key="12"
        >
          Hungary
        </p>
        <p
          key="13"
        >
          Ireland
        </p>
        <p
          key="14"
        >
          Italy
        </p>
        <p
          key="15"
        >
          Latvia
        </p>
        <p
          key="16"
        >
          Lithuania
        </p>
        <p
          key="17"
        >
          Luxembourg
        </p>
        <p
          key="18"
        >
          Malta
        </p>
        <p
          key="19"
        >
          Netherlands
        </p>
        <p
          key="20"
        >
          Poland
        </p>
        <p
          key="21"
        >
          Portugal
        </p>
        <p
          key="22"
        >
          Romania
        </p>
        <p
          key="23"
        >
          Slovakia
        </p>
        <p
          key="24"
        >
          Slovenia
        </p>
        <p
          key="25"
        >
          Spain
        </p>
        <p
          key="26"
        >
          Sweden
        </p>
        <p
          key="27"
        >
          United Kingdom
        </p>
      </div>
    </MainFrame>
  </Connect(MainFrame)>
</Provider>
`;
